tasks.go

```diff
@@ -1,125 +1,121 @@
 package main
 
-import "errors"
+import (
+	"errors"
+	"sync"
+)
 
 // A Task performs some part of the RHMAP System Dump Tool.
 type Task func() error
 
-// GetAllTasks returns a list of all tasks performed by the dump tool. It may
-// return tasks even in the presence of an error.
-// FIXME: GetAllTasks should not need to know about basepath.
-func GetAllTasks(basepath string) ([]Task, error) {
-	var (
-		tasks     []Task
-		retErrors errorList
-	)
-
-	var (
-		resources = []string{"deploymentconfigs", "pods", "services", "events"}
-		// We should only care about logs for pods, because they cover
-		// all other possible types.
-		resourcesWithLogs = []string{"pods"}
-	)
-
-	projects, err := GetProjects()
-	if err != nil {
-		return nil, err
-	}
-
-	if len(projects) == 0 {
-		return nil, errors.New("no projects visible to the currently logged in user")
-	}
+// NewError returns a Task that always return the given error.
+func NewError(err error) Task {
+	return func() error { return err }
+}
 
-	// Add tasks to fetch resource definitions.
-	definitionsTasks, err := GetResourceDefinitionsTasks(projects, resources, basepath)
-	if err != nil {
-		retErrors = append(retErrors, err)
-	}
-	tasks = append(tasks, definitionsTasks...)
+var (
+	resources = []string{"deploymentconfigs", "pods", "services", "events"}
+	// We should only care about logs for pods, because they cover all other
+	// possible types.
+	resourcesWithLogs = []string{"pods"}
+)
 
-	// Add tasks to fetch logs.
-	logsTasks, err := GetFetchLogsTasks(projects, resourcesWithLogs, basepath)
-	if err != nil {
-		retErrors = append(retErrors, err)
-	}
-	tasks = append(tasks, logsTasks...)
+// GetAllTasks returns a list of all tasks known to the dump tool.
+// FIXME: GetAllTasks should not need to know about basepath.
+func GetAllTasks(basepath string) <-chan Task {
```

Comment:
Note that the channel is only closed when all goroutines started here have terminated. And they terminate only after the tasks they produced have been received by one of the task runner goroutines from main (because the channel is unbuffered).